### ISSUE TITLE
Expose Content-Encoding to ResourceTiming

### ISSUE THREAD

**Issue by nicjansma** [w3c/resource-timing#381]
Hi,

Similar to request https://github.com/w3c/resource-timing/issues/203 to expose `Content-Type`, I would like to request exposing the `Content-Encoding` of each resource to ResourceTiming.

As we're starting to see experimentation and deployments of new content encodings such as Zstandard (`zstd`) and [compression dictionary transports](https://github.com/WICG/compression-dictionary-transport) (`zstd-d` and `br-d`), we are moving toward content being delivered from a large set of possible encodings, even to the same client on different page loads or (sub)requests to the same domain.

When the content encoded was a small set: (none), gzip and brotli, one could often infer the encoding depending on the encoded/decoded body sizes, though that generally only works if one "owns" the content (has visibility into what the size would be for each encoding type).

Having an explicit `.contentEncoding` would help with some use-cases I can think of:

* As origins and CDNs want to experiment with new Content Encodings, they could utilize phased rollouts of the above technologies and be able to segment (from RUM) the performance of clients utilizing those new technologies.
    * For example, this can help with understanding real-world uptake %, how the performance behaves for those new technologies, and debugging them
* Once fully deployed, origins and CDNs could segment browsing sessions (or A/B test) with those technologies to provide value confirmation and/or debugging via RUM.
    * For example, one could verify with RUM (along serverside logs) how often dictionaries are being used.
* Third-party RUM vendors could segment resource fetches by their encoding with certainty (rather than guessing based on file sizes)
    * For example, RUM vendors could offer insights and suggestions to enable more advanced encodings and/or dictionaries

CC @pmeenan @horo-t

**Comment by yoavweiss** [comment]
Similar to Content-Type, this probably needs to be restricted to content that is CORS-enabled or same-origin. That's not a limitation for the dictionary compression encodings (as they have similar restrictions). This might be a limitation for non-dictionary ztd/brotli.

**Comment by horo-t** [comment]
Filed chromium side bug. https://crbug.com/327941462

+CC: @jxck who is interested in this.

**Comment by Jxck** [comment]
Thanks @horo-t I'm happy to work on it !
Let me check.

**Comment by guohuideng2024** [comment]
Updated explainer can be found here:
https://github.com/guohuideng2024/resource-timing/blob/gh-pages/Explainers/Content_Encoding.md

**Comment by guohuideng2024** [comment]
I plan to filter the `contentEncoding` value before exposing it in `resource timing`. (But keep the unfiltered value in the response header for other uses), but how exactly the value should be filtered?

continuing the discussing in https://github.com/whatwg/fetch/pull/1742. 

Below are the possible values that are registered. I am putting in the following categories:

--   definitely allowed 		
br	
compress	
dcb	
dcz	
deflate	
exi	
gzip	
zstd

-- What about these?  Maybe not allow? why?

identity
pack200-gzip	
aes128gcm

-- The below are deprecated. Since we are building new feature, should we just disallow them( forcing the app to change to the up-to-date name)? Or should be be forgiving and transfer them to `compress` and `gzip`?

x-compress	Deprecated (alias for compress)	
x-gzip	Deprecated (alias for gzip)	

-- What about these? They are not registered `contendEncoding` values, should we disallow them?
br-d 
zstd-d

++++++++++++++++
A disallowed value will be exposed as `unknown`

++++++++++++++++

**Comment by pmeenan** [comment]
I'd limit it to formats a given browser supports natively (not sure if
"exi" makes that list for Chrome) and "unknown" for anything else, even if
it is registered. That way the browser can be responsible for updating
their enum as formats change.

br-d and zstd-d were replaced by dcb and dcz and should never be reported.

**Comment by guohuideng2024** [comment]
Based on what Patrick said above and some digging, I think the following could be filtered out:

`aes128gcm`: looks like this is only used for "pushing messages", not for receiving resources via fetch
`compress`, `exi`, `pack200-gzip`: not supported in chromium
`x-compress` and `x-zip`: deprecated

And the following will be supported:
`identity`

**Comment by nhelfman** [comment]
I think we must also specify what would be the value in case there is no content encoding (i.e. uncompressed payload).
It should be differentiated from the `unknown` case so we can track these separately. An example use case is when a CDN process decides to no compress the payload due to some issue (a real scenario I had). It is important to being able to identify these cases.

Seems easiest to just not set the value in that case so that when the property is read it will be JS `undefined`. This will harm supportability since it will not be possible to determine if there was no encoding or the feature is not supported / or allowed by the UA.

My suggestion is for a specific string value in that case:
`undefined` / `uncompressed` / `no-encoding` 

Anyone has a preference or other suggestions?

**Comment by yoavweiss** [comment]
[identity](https://httpwg.org/specs/rfc9110.html#field.accept-encoding:~:text=An%20%22identity%22%20token%20is%20used%20as%20a%20synonym%20for%20%22no%20encoding%22%20in%20order%20to%20communicate%20when%20no%20encoding%20is%20preferred.) seems appropriate IMO.

**Comment by guohuideng2024** [comment]
Thanks for pointing out the "identity" value meaning. I missed it. :)

And I think I could mention the "identity" value too in the `fetch` standard modification.
I am going to modify the CLs and PRs accordingly, and I will soon start reviewing process after that.

**Comment by noamr** [comment]
Do we need to decipher between "identity" and "no content-encoding header" (an empty string in `contentEncoding`)? Unlike `unknown`, those sound like the same thing to me from a client perspective?

**Comment by guohuideng2024** [comment]
@yoavweiss 

Noam point out this spec:
https://httpwg.org/specs/rfc9110.html#rfc.section.8.4
actually says that `identity` shouldn't be allowed in response header:

"Note that the coding named "identity" is reserved for its special role in [Accept-Encoding](https://httpwg.org/specs/rfc9110.html#field.accept-encoding) and thus SHOULD NOT be included."

If we cannot report `identify` value in resource timing, the client is not able to tell between:
a) no compression is used, and the server really likes to tell the client that no compression is used
(in this case, returning `identify` value is idea)
b) the server didn't update to new fetch standard so although it's using some compression, it doesn't tell the client.
(in this case, resourceTiming returning empty string)

So it appears to me that it's desirable that:

1) `identity` is allowed in fetch response, and such value is reported in resourceTiming
2) if server doesn't send `contentEncoding` value, such field is empty in resourceTiming. It means the value is missing, and CDN statistics and analysis will treat them accordingly.

==>
Then we will need to  make change to  #https://httpwg.org/specs/rfc9110.html#rfc.section.8.4?

**Comment by noamr** [comment]
> [@yoavweiss](https://github.com/yoavweiss)
> 
> Noam point out this spec: https://httpwg.org/specs/rfc9110.html#rfc.section.8.4 actually says that `identity` shouldn't be allowed in response header:
> 
> "Note that the coding named "identity" is reserved for its special role in [Accept-Encoding](https://httpwg.org/specs/rfc9110.html#field.accept-encoding) and thus SHOULD NOT be included."
> 
> If we cannot report `identify` value in resource timing, the client is not able to tell between: a) no compression is used, and the server really likes to tell the client that no compression is used (in this case, returning `identify` value is idea) b) the server didn't update to new fetch standard so although it's using some compression, it doesn't tell the client. (in this case, resourceTiming returning empty string)

What does it mean "The server didn't update to the new fetch standard"? Fetch is a client-side standards that works on top of HTTP. If the server didn't send `content-encoding` doesn't it effective means there is no compression, as in "it's sending the identity content"?

**Comment by LPardue** [comment]
Im working on other things related to HTTP content-encoding. The feedback I've received is to NOT use `identity` as label for message content that does not have a coding. 

RFC 9110 reflects the reality of use in the wild, that implementations SHOULD NOT include "identity" in Content-Encoding but they might. 

Therefore, my suggestion would be to use a label like `unencoded` (or `no-encoding` or empty string, but NOT `uncompressed` l) when there is no Content-Encoding response header. Use `identity` if there really is a `Content-Encoding: identiy` response header field.

**Comment by guohuideng2024** [comment]
Thanks for the input! So, the reality is that most of the servers just send empty string, instead of a literal string `identity`.

Let me list the possible situation here and could you guys please verify:

1) if no header is sent at all    --> return empty string
2) if empty header is sent  --> return empty string
3) Otherwise, at least one `coding` is sent. A `coding` would be transferred to `unknown` if it's not recognized. 

{gzip, apple, identity, pear } will be filtered to
{gzip, unknown, identity, unknown}
(because @LPardue  said if the server did send `identity` we can keep it?)

or we filter it to 
{gzip, unknown, unknown, unknown} with `identity` filtered too?

**Comment by noamr** [comment]
> Thanks for the input! So, the reality is that most of the servers just send empty string, instead of a literal string `identity`.
> 
> Let me list the possible situation here and could you guys please verify:
> 
> 1. if no header is sent at all    --> return empty string
> 2. if empty header is sent  --> return empty string
> 3. Otherwise, at least one `coding` is sent. A `coding` would be transferred to `unknown` if it's not recognized.
> 
> {gzip, apple, identity, pear } will be filtered to {gzip, unknown, identity, unknown} (because [@LPardue](https://github.com/LPardue) said if the server did send `identity` we can keep it?)
> 
> or we filter it to {gzip, unknown, unknown, unknown} with `identity` filtered too?

We don't want this to become a little side-channel where servers can encode information by sending mutltiple encodings over a 1-byte `fetch` or so.

So I think it should be:
- limited to maximum 2 codings. more than that becomes `multiple`. (Can be 3 if there is a common use case?)
- string separated rather than an array
- if one of them is unknown, the whole string become `unknown`
- `identity` turns to `unknown`
- No-header or empty-string turns into the empty string

Examples: `gzip`, `gzip defalate`, `br`,  `unknown`, `multiple`, "" etc.

**Comment by guohuideng2024** [comment]
Adding to what Noam said above:

1.  the filter results in a string nicely formatted, separated by a comma and a space, examples:

"" 
"gzip"   // only one coding
"gzip, deflate"   // two codings
"multiple"  // 3 or more
"unkown"  // there is something unrecognized


2. empty string sent from server along with other coding values will be ignored

below is the conversion from the raw string from server, to the string reported in resourceTiming:

", gzip"     ==>   "gzip"
", gzip,   deflate"  ==> "gzip, deflate"

3**. What if duplicated values from server?**

"gzip, Gzip"     => "gzip"?
"gzip, Gzip, deflate" => "gzip, deflate"?
"gzip, deflate, gzip" => "gzip, deflate" or "multiple"?

**Comment by noamr** [comment]
> Adding to what Noam said above:
> 3 **What if duplicated values from server**? (the server shouldn't do that but I guess we still need to define the rule here?)
> 
> "gzip, Gzip" => "gzip"? "gzip, Gzip, deflate" => "gzip, deflate"? "gzip, deflate, gzip" => "gzip, deflate" or "multiple"?

To be more concrete, these are the valid values used in this header in November 2021 in more than 100 responses (HTTP archive, mobile, out of ~1.5M:

```
COUNT | encoding
-- | --
819,032,384 | <empty>
369,685,911 | gzip
359,144,061 | br
29,541,044 | zstd
107,020 | deflate
75,470 | none
29,740 | base64
19,778 | UTF-8
13,586 | identity
11,715 | utf-8
8,247 | webp
7,788 | dcb
3,907 | text
1,955 | GZIP
1,768 | 7bit
1,323 | nosniff
925 | ISO-8859-1
873 | x-gzip
732 | null
689 | utf8
639 | None
488 | binary
454 | image/webp
408 | gzip, gzip
295 | GP5
192 | br, gzip
187 | image/jpeg
134 | application/javascript
134 | compress
107 | Stream


```

Seems like using multiple values is very rare in practice. Perhaps supplying just `unknoen`, one value or `multiple`?

**Comment by yoavweiss** [comment]
I can't think of a valid use case for multiple values. I actually now wonder what does the browser do with e.g. `br, gzip`.
The content is actually encoded with either one or the other.

Ideally, we'd report the value that was actually respected by the browser. Is that the first?

**Comment by noamr** [comment]
> I can't think of a valid use case for multiple values. I actually now wonder what does the browser do with e.g. `br, gzip`. The content is actually encoded with either one or the other.

Doesn't it mean - compress with `br`, and then compress the result with `gzip`?

**Comment by yoavweiss** [comment]
[You're right](https://httpwg.org/specs/rfc9110.html#field.content-encoding:~:text=If%20one%20or%20more%20encodings%20have%20been%20applied%20to%20a%20representation%2C%20the%20sender%20that%20applied%20the%20encodings%20MUST%20generate%20a%20Content%2DEncoding%20header%20field%20that%20lists%20the%20content%20codings%20in%20the%20order%20in%20which%20they%20were%20applied)! TIL

I still don't think it's a valid use case. A single value of "multiple" is sufficient.

**Comment by guohuideng2024** [comment]
That would be much easier to define for now.
And I will work on a CL to revise the current Chromium implementation to it.

One the other hand, I could speculate that in future "two compressions" could become more common, like a very specific compression used with a general compression technique could be beneficial. 
I am curious, when putting this to resourceTiming specification, is it possible that we leave some room so it's possible to extend in future? If yes, do we just change the specification in future (making it compatible with older version), or we just leave some room in the text we write right now?

**Comment by pmeenan** [comment]
> One the other hand, I could speculate that in future "two compressions" could become more common, like a very specific compression used with a general compression technique (Maybe some dictionary compress followed by gzip or br) could be beneficial.

I think even that would be an extreme edge case. For example, the current dictionary compressions are built directly into brotli and zstd (using `dcb` and `dcz` content encodings. Specialized compression is likely to be complete and layering them is usually a mistake or at least not helpful.

I'd prefer to keep it simple and just use `multiple` for that case, even going forward. That would provide enough information from the RUM case where someone could further dig-in and test if they aren't expecting to see it.

**Comment by guohuideng2024** [comment]
Thanks, and I think got the principle. And I think this "multiple" rule can take the highest precedence?
 i.e., as long as there is a "comma" the result is "multiple".  

Examples below are all to converted to "multiple"
", "  not ==> ""  
"gzip, gzip"  not to ==> "gzip"
"unrecognized_compression, gzip" not to "unknown"

(Because this is the simplest and the very unlikely cases don't matter much.)

**Comment by pmeenan** [comment]
Agreed. The comma means there are multiple passes at encoding, even if all of the passes are using the same encoding.

**Comment by yoavweiss** [comment]
> is it possible that we leave some room so it's possible to extend in future?

If in some unforeseeable future we'd find ourselves with a genuine case where it makes sense to apply multiple encodings, we can always decide then to add a value that represents those multiple encodings. I don't think that deciding to go with "multiple" now would have negative future-compat implications.

**Comment by pmeenan** [comment]
Sorry, which `unfiltered` contentEncoding value are you referring to? It's not an actual registered content coding so is it from a spec (fetch or resource timing) or specific to Chrome's implementation?

**Comment by guohuideng2024** [comment]
> Sorry, which `unfiltered` contentEncoding value are you referring to? It's not an actual registered content coding so is it from a spec (fetch or resource timing) or specific to Chrome's implementation?

by "unfiltered" I mean the raw "content-encoding" value that's passed via network, before the filtering. I remember I was told it can be a proprietary encoding and it is used by service worker, therefore I must keep the original value "available" while filter the value exposed by resourceTiming.

I am writing up in fetch spec so I would like to have a better understanding.   If server sends `content-encoding = my_secret_apple`, I guess the response header `content-encoding` would be `my_secret_apple` but the `content-encoding` field of `resourceTiming` should be `unknown`? 

Thanks for your patience helping me. :)

**Comment by pmeenan** [comment]
Correct. Don't change anything in fetch related to the actual handling of the content-encoding itself. The only place where the filtering should be applied is when the value is being exposed/copied to resource timing.

**Comment by guohuideng2024** [comment]
@annevk  and I had discussion regarding the `unknown` value. We like to use a literal string that cannot be a future legit content-encoding value to avoid possible (although unlikely) conflict in future.

After discussing with itef, we failed to reserve `unknown` value. However as @annevk noticed, that Roy Fielding pointed out `@unknown` cannot be registered as `content encoding` value. So we are going to use `@unknown` instead of `unknown`.

I will wait for couple of days to see if there is any objections before proceeding. Thanks!

**Comment by guohuideng2024** [comment]
Update: I found a few alternatives, like `?unknown` and `[unknown]`, `(unknown)`. (Possibly more candidates can be reasonable, see below for details)

Just like to confirm we are going use `@unknown`  and @annevk and I will continue to discuss at https://github.com/whatwg/fetch/pull/1796

If anyone thinks an alternative is better, please let us know, preferably at https://github.com/whatwg/fetch/pull/1796
And if we change something I will update here.


++++++++++++++++++++++++++++++++++++
I found that the http spec actually defines the syntax restriction of "content-coding" parameter.
Link to the http spec: https://www.rfc-editor.org/rfc/rfc2616
The "content-coding" is a "token", and the "token" is defined as the following:

       token          = 1*<any CHAR except CTLs or separators>
       separators     = "(" | ")" | "<" | ">" | "@"
                      | "," | ";" | ":" | "\" | <">
                      | "/" | "[" | "]" | "?" | "="
                      | "{" | "}" | SP | HT
// where:
       CTL            = <any US-ASCII control character
                        (octets 0 - 31) and DEL (127)>
       SP             = <US-ASCII SP, space (32)>
       HT             = <US-ASCII HT, horizontal-tab (9)>

**Comment by annevk** [comment]
I wouldn't mind `:unknown` (after the HTTP/2 syntax for pseudo-fields), but `@unknown` is also still okay.
